src/main/java/io/openmarket/marketplace/MarketPlaceServiceHandler.java

```diff
@@ -0,0 +1,179 @@
+                    .build()
+            );
+        }
+
+        return GetOrgItemsResult.newBuilder().addAllItems(result).build();
+    }
+
+    public CheckOutResult checkout(String uesrId, CheckOutRequest request) {
+        List<ItemGrpc> items = request.getItemsList();
+        List<Integer> itemCounts = request.getCountList();
+        double total = 0.0;
+        boolean done = false;
+        if (items.size() != itemCounts.size()  || items.isEmpty()
+        || request.getCurrencyId().isEmpty() || request.getFromOrg().isEmpty())
+            throw new IllegalArgumentException("Invalid order");
+
+
+        log.info("Checkout received from {} buying from {}",
+                uesrId, request.getFromOrg());
+
+        //puts item on hold, not finalized until transcation is processed correctly
+       CheckOutResult result = updateItemDB(items, itemCounts, null);
+
+       if (result.getUnprocessedItemCount() == items.size()) return result;
+
+       Set<ItemGrpc> unprocessed = new HashSet<>(result.getUnprocessedItemList());
+       for (int i = 0; i < items.size(); i++) {
+           ItemGrpc item = items.get(i);
+           int count = itemCounts.get(i);
+
+           if (!unprocessed.contains(item)) {
+               total += this.itemDao.load(item.getItemId()).get().getItemPrice() * count;
+           }
+       }
+
+        String transactionId = transactionServiceHandler.createPayment(uesrId,
+                TransactionProto.PaymentRequest.newBuilder()
+                        .setType(TransactionProto.PaymentRequest.Type.TRANSFER)
+                        .setRecipientId(request.getFromOrg())
+                        .setMoneyAmount(TransactionProto.MoneyAmount.newBuilder()
+                                .setCurrencyId(request.getCurrencyId())
+                                .setAmount(total)
+                                .build())
+                        .build()
+        );
+
+        if(!checkTransactionStatus(transactionId)) {
+            log.error("Transaction timed out");
+            //revert item from on hold to available again
+            updateItemDB(items, itemCounts, unprocessed);
+            return CheckOutResult.newBuilder()
+                    .setCheckoutStatus(CheckOutResult.Status.FAIL_TRANSACTION_TIME_OUT)
+                    .addAllUnprocessedItem(items)
+                    .build();
+        }
+        log.info("Checkout payment verifield from {}", uesrId);
+
+        return result;
+    }
+
+    private boolean checkTransactionStatus(String transactionId) {
+        for (int i = 0; i < CHECKOUT_TIMEOUT; i++) {
+            TransactionStatus paymentStatus = transactionServiceHandler.getTransactionStatus(transactionId);
+            if (paymentStatus.equals(TransactionStatus.COMPLETED)){
+                return true;
+            } else if (paymentStatus.equals(TransactionStatus.ERROR)) {
+                return false;
+            }
+            try{ Thread.sleep(2000); }
```

Comment:
Shouldn't this be 1000?

Reply:
Fixed in feature-getsimilar pull request